### PR TITLE
Kernel: Make Process::FileDescriptions::allocate return KResultOr<int>

### PR DESCRIPTION
**Pull request by bgianfo**
- **Kernel: Avoid file descriptor leak in Process::sys$socketpair on error**

    Previously it was possible to leak the file descriptor if we error out
    after allocating the first descriptor. Now we perform both fd
    allocations back to back so we can handle the potential error when
    processing the second fd allocation.

- **Kernel: Track allocated FileDescriptionAndFlag elements in each Process**

    The way the Process::FileDescriptions::allocate() API works today means
    that two callers who allocate back to back without associating a
    FileDescription with the allocated FD, will receive the same FD and thus
    one will stomp over the other.

    Naively tracking which FileDescriptions are allocated and moving onto
    the next would introduce other bugs however, as now if you "allocate"
    a fd and then return early further down the control flow of the syscall
    you would leak that fd.

    This change modifies this behavior by tracking which descriptions are
    allocated and then having an RAII type to "deallocate" the fd if the
    association is not setup the end of it's scope.

- **Kernel: Make Process::FileDescriptions::allocate return KResultOr<int>**

    Modernize more error checking by utilizing KResultOr.

- **Kernel: Remove unused fd allocation from Process::sys$connect(..)**

### CODE REVIEW

File: Kernel/Process.cpp [hunk 1]
```diff
@@ -465,14 +465,14 @@ size_t Process::FileDescriptions::open_count() const
...
-int Process::FileDescriptions::allocate(int first_candidate_fd)
+KResultOr<int> Process::FileDescriptions::allocate(int first_candidate_fd)
 {
     ScopedSpinLock lock(m_fds_lock);
     for (size_t i = first_candidate_fd; i < max_open(); ++i) {
         if (!m_fds_metadatas[i])
             return i;
     }
-    return -EMFILE;
+    return KResult(EMFILE);
```
Comment (gunnarbeutner) [inline Kernel/Process.cpp:475]: The cast shouldn't be necessary.

File: Kernel/Syscalls/socket.cpp [hunk 2]
```diff
@@ -412,21 +412,23 @@ KResultOr<FlatPtr> Process::sys$socketpair(Userspace<const Syscall::SC_socketpai
         return result.error();
     auto pair = result.value();
 
-    int fds[2];
+    auto fd0_or_error = m_fds.allocate();
+    if (fd0_or_error.is_error())
+        return fd0_or_error.error();
     auto fd1_or_error = m_fds.allocate();
```
Comment (gunnarbeutner) [inline Kernel/Syscalls/socket.cpp]: Wouldn't that give us the same FD for both `allocate()` calls - considering that `allocate()` does not actually "allocate" the FD.

Comment (awesomekling) [reply Kernel/Syscalls/socket.cpp]: Indeed. It seems to me that `allocate()` should return some sort of temporary holder of an allocated FD. This holder would release the allocation on destruction, but the FD could also be moved out of it.

Comment (bgianfo) [reply Kernel/Syscalls/socket.cpp]: Good points, I cooked up a scoped type to handle the allocation / de-allocation and added tracking...

File: Kernel/Process.cpp (hunk 1 above)
Comment (bgianfo) [reply Kernel/Process.cpp]: Yup, fixed.

**Pull request merged**